packages/expo-router/src/rsc/rsc-renderer.ts

```diff
@@ -0,0 +1,264 @@
+    };
+    return runWithRenderStore(renderStore, async () => {
+      const elements = await renderEntries(input, {
+        searchParams,
+        buildConfig,
+      });
+      if (elements === null) {
+        const err = new Error('No function component found at: ' + input);
+        (err as any).statusCode = 404;
+        throw err;
+      }
+      if (Object.keys(elements).some((key) => key.startsWith('_'))) {
+        throw new Error('"_" prefix is reserved');
+      }
+      return renderToReadableStream(elements, bundlerConfig);
+    });
+  };
+
+  const renderWithContextWithAction = async (
+    context: Record<string, unknown> | undefined,
+    actionFn: (...args: unknown[]) => unknown,
+    actionArgs: unknown[]
+  ) => {
+    let elementsPromise: Promise<Record<string, ReactNode>> = Promise.resolve({});
+    let rendered = false;
+    const renderStore = {
+      context: context || {},
+      rerender: async (input: string, searchParams = new URLSearchParams()) => {
+        if (rendered) {
+          throw new Error('already rendered');
+        }
+        elementsPromise = Promise.all([
+          elementsPromise,
+          renderEntries(input, { searchParams, buildConfig }),
+        ]).then(([oldElements, newElements]) => ({
+          ...oldElements,
+          // FIXME we should actually check if newElements is null and send an error
+          ...newElements,
+        }));
+      },
+    };
+    return runWithRenderStore(renderStore, async () => {
+      const actionValue = await actionFn(...actionArgs);
+      const elements = await elementsPromise;
+      rendered = true;
+      if (Object.keys(elements).some((key) => key.startsWith('_'))) {
+        throw new Error('"_" prefix is reserved');
+      }
+      return renderToReadableStream({ ...elements, _value: actionValue }, bundlerConfig);
+    });
+  };
+
+  if (method === 'POST') {
+    // TODO(Bacon): Fix Server action ID generation
+    const rsfId = decodeURIComponent(input);
+    let args: unknown[] = [];
+    let bodyStr = '';
+    if (body) {
+      bodyStr = await streamToString(body);
+    }
+    if (typeof contentType === 'string' && contentType.startsWith('multipart/form-data')) {
+      // XXX This doesn't support streaming unlike busboy
+      const formData = parseFormData(bodyStr, contentType);
+      args = await decodeReply(formData, bundlerConfig);
+    } else if (bodyStr) {
+      args = await decodeReply(bodyStr, bundlerConfig);
+    }
+    const [, name] = rsfId.split('#') as [string, string];
+    // xxxx#greet
+    if (!getServerReference(rsfId)) {
+      throw new Error(`Server action not found: "${rsfId}". ${getDebugDescription()}`);
+    }
+    const mod: any = getServerReference(rsfId);
+
+    const fn = name ? mod[name] || mod : mod;
+    return renderWithContextWithAction(context, fn, args);
+  }
+
+  // method === 'GET'
+  return renderWithContext(context, input, searchParams);
+}
+
+// TODO is this correct? better to use a library?
```

Comment:
Let's check what Undici does, maybe we can use some parts of Undici to get this done?

Reply:
This part of the forked code and is hopefully temporary. Likely will be part of React in the future.